pkg/endpointmanager/manager.go

```diff
@@ -399,10 +405,12 @@ func AddEndpoint(owner endpoint.Owner, ep *endpoint.Endpoint, reason string) (er
 	}
 
 	if build {
-		if err := ep.RegenerateWait(owner, reason); err != nil {
-			Remove(ep)
-			return err
-		}
+		go func() {
+			if err := ep.RegenerateWait(owner, reason); err != nil {
```

Comment:
Callers could not rely on this as it was dependent on the state already. We only have two callers to `AddEndpoint()`:

1. `createEndpoint()`: When new endpoints are inserted
2. `LaunchAsEndpoint()` when the health endpoint is created.

We should move to `RegenerateWait()` into the callers:

* For `LaunchAsEndpoint()` it should run as a controller to retry on error and also have an interval for rebuilds such as 15 minutes so we have a fallback

* For `createEndpoint()` it should also start a controller to keep retrying to build the endpoint. We can do an interval of 15 minutes as well. When `createEndpoint()` is in `SyncBuildEndpoint` mode then it should first do the sync build and then start the controller.

The controller must be removed on endpoint leave.

Once we have that we can follow-up and use the controller for all endpoint regenerations and we always have retries on build failures.

Reply:
I wasn't sure whether that was in the scope of this PR or not. I wanted to minimize the existing changes in behavior in the cilium-agent during one of the most critical parts of code (creation of endpoints) here, so I made the regeneration of the endpoint asynchronous, but still removed it upon failure if the first build failed, which preserves existing behavior. I did this because this needs to be backported to the v1.4 branch. @tgraf do you want this change to be made as part of this PR? I'm worried that we may change behavior in a way we don't expect with that sort of a change, which may be better to do as part of the work for v1.5. I may be overly cautious here, though.